### PR TITLE
Add parsing of hex literals in strings

### PR DESCRIPTION
**Pull request by markuspf**
With this patch it is possible to give character values by using
the common '\xHH' escape where H is a hexadecimal digit.

If this is accepted, I wonder whether there is also an interest for python's `\u`? It might not be as easy as this one as `\u` parses to more than one byte.

I think this might break existing code though, because 

```
gap> "\xff";
"xff"
```

### CODE REVIEW

**Comment by markuspf** [comment]
Apparently gcc 4.x doesn't like inline function definitions that are in headers and not guarded. Well, I'll move the function to `system.c` then

**Comment by rbehrends** [comment]
Not sure what you mean by "not guarded", but may the issue simply be that the function needs to be "static inline" instead of just "inline"?

**Comment by markuspf** [comment]
m( of course. Thanks for pointing that out.

**Comment by stevelinton** [comment]
Happy with this. It would be nice to see a comprehensive approach to dealing with Unicode strings (for instance one which allowed them to be viewed as a list of unicode characters) and then it might make sense to add some syntactic support for them, but I think it would be good to have the grand design first.

**Comment by markuspf** [comment]
This PR also is missing documentation changes and tests. I will add those
soon.

**Comment by ChrisJefferson** [comment]
While it doesn't directly fit in this patch, is there any reason not to make '\a' for all characters 'a' which do not have a special meaning an error, rather than just making the '\' get ignored.

**Comment by markuspf** [comment]
I wondered that too. It should be easy to add either as a separate PR if people think it's sensible.

**Comment by frankluebeck** [comment]
I don't remember why hex characters in string literals were not introduced long ago. At least now I cannot see a reason against it. I would consider the slight incompatibility of `"\x"` acceptable.

That the `\` is ignored if it is not followed by a special character was already in very early versions of GAP. Nevertheless, I guess that it will not disturb many people if this was changed. This had the advantage that afterwards  one could introduce further special characters without breaking backward compatibility again.

GAP strings are really sequences of bytes. Introducing `\uXXXX` for unicode characters does not really make sense (because a GAP string has no encoding context). 
Note that there is functionality for unicode strings:
```
?Unicode Strings
```
The `Unicode` function was originally called `U` but I changed it after people complained about a new one letter function.

**Comment by stevelinton** [comment]
@frankluebeck great. We should add some cross-references from the strings section of the reference manual to the GAPDoc manual.  Is there are a case for splitting GAPDoc into several packages? XML, Unicode, help viewers, help compiler?

Do we need kernel support for (for instance) reading large UTF-8 files into a Unicode string?

**Comment by fingolfin** [comment]
@stevelinton These are all interesting points, but perhaps for the mailing list, not on this PR, which will (hopefully) soon be merged, and then the discussion is "lost".

File: doc/ref/string.xml [hunk 1]
```diff
@@ -150,7 +150,9 @@ doublequotes of a string literal to specify characters.
 They consist of two characters.
 The first is a backslash <C>\</C>. The second may be any character.
 If it is an octal digit (from <C>0</C> to <C>7</C>) there must be two more
-such digits. The meaning is given in the following list
+such digits.
+If it is the character <C>x</C>, then there must be two hexadecimal
+digits. The meaning is given in the following list
```
Comment (fingolfin) [inline doc/ref/string.xml]: This sounds weird, in particular the bit saying "then there must be two hexadecimal such digits", which seems to be missing something. Finally, the sentence "They consist of two characters." was already before this patch quite misleading. How about this:
```
There are a number of <E>special character sequences</E> that can be used
 between the singlequotes of a character literal or between the
 doublequotes of a string literal to specify characters.
They consists of a backslash <C>\</C>, followed by a second character indicating the type of
special character sequence, and, depending on this type, possibly some more characters.

The following special character sequencs are currently defined. Any other sequence starting with
a backslash results in an error.
```

File: doc/ref/string.xml [hunk 2]
```diff
@@ -234,6 +236,17 @@ such digits. The meaning is given in the following list
     in &GAP;.
 </Item>
```
Comment (fingolfin) [inline doc/ref/string.xml]: If you follow my suggestion, you could augment the octal digits part by adding the `(from <C>0</C> to <C>7</C>)` bit as explanantion for the "octal digits"

File: src/system.h [hunk 3]
```diff
@@ -630,6 +630,16 @@ extern UInt SyTimeChildrenSys ( void );
 */
 #define IsDigit(ch)     (isdigit((unsigned int)ch))
 
+
+/****************************************************************************
+**
+*F  IsHexDigit( <ch> ) . . . . . . . . . . . . . . .  is a character a digit
+**
+**  'IsDigit' returns 1 if its character argument is a digit from the ranges
+**  '0..9' or 'a..f' and 0 otherwise.
```
Comment (fingolfin) [inline src/system.h]: This leaves out A..F

**Comment by fingolfin** [comment]
I think this is a good idea, thanks. Just have some minor quibbles with the documentation.

**Comment by markuspf** [comment]
Well this is fun:

On Sun, Feb 14, 2016 at 11:57:38PM -0800, Christopher Jefferson wrote:
> While it doesn't directly fit in this patch, is there any reason not to make '\
> a' for all characters 'a' which do not have a special meaning an error, rather
> than just making the '\' get ignored.

I tried patching this in and it turns out that in the library and packages there
are a few places that have "\!", "\*", "\-", "\#" and others. While I suspect
that these are acutally bugs, the fallout from such a patch might be a tad
larger than I thought.

I'll make a separate PR for this...

**Comment by markuspf** [comment]
I should probably pull the refactoring from #619 into this PR before we merge it.

One small issue I see is that the error message is currently less specific. That could be addressed though if people are worried about it.

**Comment by markuspf** [comment]
I updated this PR in the following way
 * Change the escape sequence for hexadecimal values in strings and characters to `\0xHH`
 * Include the refactoring done in #619
 * Display a syntax warning if a `\\` is followed by anything that is not a valid escape sequence.

**Comment by markuspf** [comment]
I have fixed `qaos` and sent an email to Thomas Breuer about ctbllib.

**Comment by frankluebeck** [comment]
Looks mostly fine to me. Using \0x?? for the hex characters is similar to the syntax in other languages. And it has the advantage that there is no backward compatibility problem. Of course, the refactoring is also sensible. 

But as discussed in the thread on "Stricter string escape" I'm not happy with the  warning if a backslash is used with non-special characters (and the corresponding change of the documentation). This change seems purely cosmetic and it breaks without real need a behaviour that is documented since 30 years (and makes changes in several places necessary).

**Comment by markuspf** [comment]
Is there any preference regarding whether we just revert to the old behaviour with respect to `\\` or display a warning if `\\` is followed by a letter that is not in a valid sequence?

I'd quite like to move this PR forward.

**Comment by stevelinton** [comment]
> On 29 Feb 2016, at 11:16, Markus Pfeiffer <notifications@github.com> wrote:
> 
> Is there any preference regarding whether we just revert to the old behaviour with respect to \\ or display a warning if \\ is followed by a letter that is not in a valid sequence?
> 
> I'd quite like to move this PR forward.
> 
> —
> Reply to this email directly or view it on GitHub.
> 

Mild preference for the warning.

Steve

**Comment by markuspf** [comment]
Now updated to only warn if a backslash is followed by a letter.

Tests fail because of minor errors in `qaos` (for which there is a new version available that does not have this error) and `ctbllib` (which I reported to Thomas Breuer).

**Comment by fingolfin** [comment]
What is the status of this? It seems like it was basically done (though travis failed), and nobody really had complaints left?

@markuspf Perhaps you can rebase it, so that codecov gets run; I'd be happy to review this once more afterwards, too.

**Comment by alex-konovalov** [comment]
IIRC, the new version of qaos contained a fix for the problem addresses by this PR, but ctbllib will fail the `bugfix.tst` test:
```
testing: /home/travis/build/gap-system/gap/tst/teststandard/bugfix.tst
rev: ########> Diff in /home/travis/build/gap-system/gap/tst/teststandard/bugfix.ts\
t:2884
# Input is:
if LoadPackage("ctbllib", false) <> fail then
     if Irr( CharacterTable( "WeylD", 4 ) )[1] <>
          [ 3, -1, 3, -1, 1, -1, 3, -1, -1, 0, 0, -1, 1 ] then
       Print( "problem with Irr( CharacterTable( \"WeylD\", 4 ) )[1]\n" );
     fi;
   fi;
# Expected output:
# But found:
Syntax warning: Alphabet letter after \ in /home/travis/build/gap-system/gap/p\
\
kg/ctbllib/data/ctgeneri.tbl:915
"in this paper (the characters $\Theta_l$ and $\Lambda_u$ on the classes\n",
                                ^
Syntax warning: Alphabet letter after \ in /home/travis/build/gap-system/gap/p\
\
kg/ctbllib/data/ctgeneri.tbl:915
"in this paper (the characters $\Theta_l$ and $\Lambda_u$ on the classes\n",
                                               ^
########
```

**Comment by fingolfin** [comment]
So, how about completly disabling the warning for now (leave it in, but commented out)? We can re-evaluate enabling it in the future, when/if `ctbllib` gets changed. In the meantime, I don't think we should hold up a useful feature (support for hex literals) because of this...

**Comment by markuspf** [comment]
@finglfin there you go. I disabled the syntax warning for now, though I have to admit I am uncomfortable with bugwards-compatibility hacks.

**Comment by frankluebeck** [comment]
I have not changed my opinion compared to my comment on Feb 26. The example from ctblib mentioned above even demonstrates that GAPs (documented!) behaviour  can be  useful.

File: doc/ref/string.xml [hunk 4]
```diff
@@ -227,17 +227,30 @@ such digits. The meaning is given in the following list
...
+    of <C>"01234567"</C>.
     This is translated to the character corresponding to the number
     <C>X * 64 + Y * 8 + Z modulo 256</C>.
     This can be used to specify and store arbitrary binary data as a string
     in &GAP;.
 </Item>
 <Mark>
+<Index><C>\0xYZ</C></Index>
+<Index>hexadecimal character codes</Index>
+<C>\xYZ</C></Mark>
```
Comment (fingolfin) [inline doc/ref/string.xml]: This still says `\xYZ`, should be `\0xYZ`.

File: doc/ref/string.xml [hunk 5]
```diff
@@ -227,17 +227,30 @@ such digits. The meaning is given in the following list
...
+    <C>15</C>. This is translated to the character corresponding
+    to the number <C>Y*16 + Z</C>.
+</Item>
+<Mark>
 <Index>escaping non-special characters</Index>
 other</Mark>
 <Item>
-    For any other character the backslash is simply ignored.
+  For any other character the backslash is ignored. If the character
+  is a letter, that is one of <C>a..zA..Z</C>, then a warning is displayed.
```
Comment (fingolfin) [inline doc/ref/string.xml]: Same as above: If we can't agree on changing the documented and actual behaviour, then this last sentence is controversial. In that case, we should drop it, but can keep the rest of this change.

File: doc/ref/string.xml [hunk 6]
```diff
@@ -147,10 +147,10 @@ next line
...
 between the singlequotes of a character literal or between the
 doublequotes of a string literal to specify characters. 
-They consist of two characters.
-The first is a backslash <C>\</C>. The second may be any character.
-If it is an octal digit (from <C>0</C> to <C>7</C>) there must be two more
-such digits. The meaning is given in the following list
+They consist of a backslash <C>\</C> followed by a second character
+indicating the type of special character sequence, and possibly more characters.
+The following special character sequences are currently defined. Any other sequence
+starting with a backslash results in an error.
```
Comment (fingolfin) [inline doc/ref/string.xml]: If we can't agree on changing the documented and actual behaviour, then this last sentence is controversial. In that case, we should drop it, but can keep the rest of this change.

File: lib/helpbase.gi [hunk 7]
```diff
@@ -416,7 +416,7 @@ InstallGlobalFunction(SIMPLE_STRING, function(str)
 "efghijklmnopqrstuvwxyz[\000]^_\000abcdefghijklmnopqrstuvwxyz{ }~",
 "\177\200\201\202",
 "\203\204\205\206\207\210\211\212\213\214\215\216\217\220\221\222\223\224\225",
-"\226\227\230\231\232\233\234\235\236\237\238",
+"\226\227\230\231\232\233\234\235\236\237\240",
```
Comment (fingolfin) [inline lib/helpbase.gi]: What is the purpose of this change?

File: src/scanner.c [hunk 8]
```diff
@@ -1810,6 +1810,81 @@ void GetNumber ( UInt StartingStatus )
...
+    } else {
+        SyntaxError("Expecting hexadecimal escape, or two more octal digits");
+    }
+  } else if ( *TLS(In) >= '1' && *TLS(In) <= '7' ) {
+    /* escaped three digit octal numbers are allowed in input */
+    c = 64 * (*TLS(In) - '0');
+    GET_CHAR();
+    c += GetOctalDigits();
+  } else {
+      /* This warning is currently disabled for backwards compatibility */
```
Comment (fingolfin) [inline src/scanner.c]: Perhaps elaborate a bit, and/or reference this PR. E.g.:
"This warning is currently disabled for backwards compatibility: It turns out there are some packages which still rely on this."

File: src/system.h [hunk 9]
```diff
@@ -631,6 +631,16 @@ extern UInt SyTimeChildrenSys ( void );
...
+
+/****************************************************************************
+**
+*F  IsHexDigit( <ch> ) . . . . . . . . . . . . . . .  is a character a digit
+**
+**  'IsDigit' returns 1 if its character argument is a digit from the ranges
+**  '0..9', 'A..F', or 'a..f' and 0 otherwise.
+*/
+#define IsHexDigit(ch)     (isxdigit((unsigned int)ch))
+
```
Comment (fingolfin) [inline src/system.h]: OK. Technically, there might be systems out there which don't have `isxdigit`, but we'll just deal with them if we ever encounter them. So this is fine.

File: src/system.h [hunk 10]
```diff
@@ -643,6 +653,21 @@ extern UInt SyTimeChildrenSys ( void );
...
+*/
+static inline Char CharHexDigit( const Char ch ) {
+    if (ch >= 'a') {
+        return (ch - 'a' + 10);
+    } else if (ch >= 'A') {
+        return (ch - 'A' + 10);
+    } else {
+        return (ch - '0');
+    }
+};
```
Comment (fingolfin) [inline src/system.h]: This is only used in one place... so unless we forsee use for it, why not just move it to `scanner.c` ? If something else ever needs it, we can still move it to a header (and then contemplate whether it needs extra safe guards or not etc.)

File: tst/testinstall/strings.tst [hunk 11]
```diff
@@ -115,6 +121,12 @@ gap> PrintString(x);
 "'a'"
 gap> String(x);
 "'a'"
+gap> x:='\0x42';
+'B'
+gap> x:='\0xFF';
+'\377'
+gap> x:='\0xab';
+'\253'
```
Comment (fingolfin) [inline tst/testinstall/strings.tst]: These test cases only cover positives, but not malformed inputs, which we should also test. E.g. `\0yAB`, `\0X12`, `\090`, `\0a0`, `\0x0`,  `\009`, `\00a`, `\00x`, `\0x1g`, ..

File: lib/helpbase.gi (hunk 7 above)
Comment (markuspf) [inline lib/helpbase.gi]: \238 is not a valid octal escape, whereas \240 is (for the same number, if one were to allow overflows in the octal digits).

File: src/system.h (hunk 9 above)
Comment (markuspf) [inline src/system.h]: `isxdigit` is in ISO C90, I thought we're doing C99 even? Am I mixing something up?

**Comment by markuspf** [comment]
I adapted the PR according to @fingolfin's comments. I should probably do some history rearrangement before we merge this (If everything is ok).

File: lib/helpbase.gi (hunk 7 above)
Comment (fingolfin) [inline lib/helpbase.gi]: ahhhh of course! *facepalm*

**Comment by fingolfin** [comment]
Looks good to me now, thanks. If you want to cleanup the history, go ahead. Other than that, i think it can be merged now.
